### PR TITLE
Fix for s3 TF state bucket logs loop - issue 1333

### PR DESCRIPTION
**Pull request by ina-stoyanova**
This PR is aiming to fix this issue: https://github.com/gruntwork-io/terragrunt/issues/1333

- [x] Fix related bug (pre-existing) to this issue - [more details can be found here.](https://github.com/gruntwork-io/terragrunt/pull/1399) Thanks to @yorinasub17 & @zackproser 
- [x] Decide what to do with `skip_bucket_access_logging` - do we deprecate or fix forward & replace with `enable_bucket_access_logging`)
    - see below in comments 
- [x] Implement `enable_bucket_access_logging` as bool to flip logic around - e.g. if not provided, by default the logging will be turned off for the bucket. _**Thanks to @infraredgirl**_
- [x] implement `enable_bucket_access_logging` as a string - if provided, we enable access logging for the bucket & use the string param as a name for the bucket; if not provided - disable access logging by default _**Thanks to @infraredgirl**_
- [x] implement create bucket for access logs if name is provided, but bucket does not exist yet
- ~[x] refactor for reausability the s3 create & set access methods~
- [x] refactor to avoid duplication `createS3BucketIfNecessary` & `CreateLogsS3BucketIfNecessary` - potentially we can pass function call as a param?
- [x] update/ add to automated tests
- [x] update logging so it's clear that a setting has been turned on:
      - `EnableVersioningForS3Bucket` (e.g. enabling..... -> enabled.....)
      - `EnableSSEForS3BucketWide`
      - `EnableAccessLoggingForS3BucketWide`
      - `EnablePublicAccessBlockingForS3Bucket`
      - `configureBucketAccessLoggingAcl`

### CODE REVIEW

**Comment by ina-stoyanova** [comment]
[last-terragrunt-apply-output.txt](https://github.com/gruntwork-io/terragrunt/files/5430691/last-terragrunt-apply-output.txt)
☝️ this is the output of the code ran at the last commit state. 

All the logs are expected for the following terragrunt config:
```
# stage/terragrunt.hcl
remote_state {

  backend = "s3"
  generate = {
    path      = "backend.tf"
    if_exists = "overwrite_terragrunt"
  }

  config = {
    bucket = "ina-1"
    accesslogging_bucket_name = "ina-1-logs"
    key = "${path_relative_to_include()}/terraform.tfstate"
    region         = "eu-west-2"
    encrypt        = true
    dynamodb_table = "ina-1"
  }
}
```

**Comment by ina-stoyanova** [comment]
Next steps to pick up:
- confirm we can still see the same behaviour as the logs above show
- confirm if param is not provided that S3 bucket for logs is not created
- confirm if we need to sanitise input for name of s3 bucket
- confirm what happens to `skip...` parameter
- refactor duplication between createS3 bucket methods (for logs, and for standard state bucket)

**Comment by ina-stoyanova** [comment]
As part of this PR, the default behaviour of terragrunt's functionality to create the TF State S3 bucket will change. 

**- previous implementation:** create TF State S3 bucket with server access logging enabled & set to the origin (same) S3 bucket. _For more details, please read through https://github.com/gruntwork-io/terragrunt/issues/1333, demonstrating why this is undesirable behaviour._  

**- new implementation with this PR"** create TF State S3 bucket with server access logging disabled. Terragrunt users will be able to enable server access logging for the TF State S3 bucket by providing a name for the desired logs target bucket. This can be done using the `accesslogging_bucket_name <string>` parameter in the `terragrunt.hcl` configuration of their implementation. 

In more detail:
- The `skip_bucket_accesslogging` terragrunt option is being deprecated. It's no longer going to take effect on the S3 bucket for TF state creation and logging configuration. 
- A new terragrunt configuration option is added: `accesslogging_bucket_name <string>`
    - states in the `terragrunt.hcl` configuration file
         - not present -> server access logging disabled by default for TF State S3 bucket
         - empty -> server access logging disabled by default for TF State S3 bucket
         - present (valid string) -> the server access logging for the TF State S3 bucket will be enabled; the target logging bucket will be the S3 bucket with name as specified by the configuration option value `accesslogging_bucket_name` 
              - if bucket with such name does not exist, the user will be prompted to create the S3 bucket.


Deprecating `skip_bucket_accesslogging`:

_We've got 3 options identified as to how to proceed. After chatting to @brikis98 we've decided it's best to go with a combination of **1** & **2**. Option 3 is the most ideal scenario one, and feels like it's outside of the scope of this PR currently. As a way to see what's the engagement/appetite for option 3, here's an issue raised on the Terragrunt repo. _

1. basic option - remove the flag & update the docs & release notes 
2. we detect the param - show error & provide guidance 
3. warn people of existing bug & point them to fix it/or we automatically fix it for them (existing buckets)

**Comment by ina-stoyanova** [comment]
Tests have been updated very little. The reasoning is, the important thing is that the terraform state is initialised using s3 backend. I couldn't find any other examples of the other config options being tested. 

For the `accesslogging_bucket_name` config option, it feels like the effort to automate the testing of it will be more than the value of it would actually bring. I'm happy to take guidance if anyone feels strongly we should have the tests cover this explicitly. If that's the direction we take, I'd suggest doing it separately, so we can also test the rest of the properties alongside with this one, which would increase the test confidence in what the `terraform init` will affect if run with terragrunt. 

I have tested manually the following cases:
- not providing `accesslogging_bucket_name` => ✅
- providing the value `accesslogging_bucket_name` => ✅ 
- providing incorrect S3 name for the bucket => ✅ 
- providing `accesslogging_bucket_name` as incorrect type of input => ✅ 
- keeping the `skip_bucket_accesslogging` => ✅ 
- providing value (bool) for the `skip_bucket_accesslogging` => ✅

File: remote/remote_state_s3.go
Comment (yorinasub17) [inline remote/remote_state_s3.go:346]: This change of moving `DoesS3BucketExist` check to `CreateS3Bucket` makes me nervous because it is a fairly significant change in behavior.

Before, terragrunt never attempted to "migrate" an existing bucket to match the configurations and defaults that you have set. This means that even if we update bucket parameters across releases, and if the user upgrades terragrunt versions, the s3 bucket remained untouched. This is somewhat surprising and confusing behavior for folks who are used to terraform, but has a few advantages for us. Off the top of my head:

- It is simpler to maintain, as we didn't really have to worry about backwards compatibility in the s3 bucket creation code: we always started from scratch with a new bucket.

- It doesn't have the behavior of modifying the state bucket during a `plan` or `validate` stage, which can be surprising to a user attempting to do a "read-only" task (NOTE: creating the bucket also happens during these stages, which is arguably just as surprising, but in the create case there is an interactive prompt to confirm the creation that softens the impact).

There is potential that this change can break a lot of existing users, and we also should think through the implications of applying some of the updates that this will do. From the looks of it, this change appears to be orthogonal to the access logging change, so I think this specific update should be in a separate PR where we can discuss the implications of this update.

File: remote/remote_state_s3.go
Comment (yorinasub17) [inline remote/remote_state_s3.go:423]:
```suggestion
		terragruntOptions.Logger.Printf("___WARNING___: The terragrunt configuration option 'skip_bucket_accesslogging' is now deprecated. Access logging for the state bucket %s is disabled by default. To enable access logging for bucket %s, please provide property `accesslogging_bucket_name` in the terragrunt config file. For more details, please refer to the Terragrunt documentation.", config.remoteStateConfigS3.Bucket, config.remoteStateConfigS3.Bucket)
```

File: remote/remote_state_s3.go
Comment (yorinasub17) [inline remote/remote_state_s3.go:428]: This message needs to be updated.

File: remote/remote_state_s3.go
Comment (yorinasub17) [inline remote/remote_state_s3.go:426]: There should be a log message here indicating that terragrunt is going to create the access logging bucket and enable it.

**Review by yorinasub17 — COMMENTED** [review]
The updated implementation LGTM. The deprecation policy we've implemented here makes sense and is approachable.

I did have a concern about one of the changes that was implemented here, which could use a second pair of eyes to confirm if my theory is correct and if my concerns are valid. See the comment below.

Regarding testing, it would be good to have a sanity check regression integration test that makes sure terragrunt sets up access logging to the right bucket. It doesn't have to be comprehensive/complicated where it validates if access logs actually get generated, but an existence and config sanity check to make sure the API calls are correct and happen in the right order would be useful.

File: remote/remote_state_s3.go
Comment (ina-stoyanova) [reply remote/remote_state_s3.go:346]: Thank @yorinasub17 This is really useful insight. Here's some of my thoughts, however, based on the concerns you've got, it might be easier to chat through this on a call so I can understand better _(I suspect I'm missing some context & understanding)_

- the `CreateS3Bucket` func has only been called from 2 places, one of which is the new introduction for creating logs bucket, and the other one being `CreateS3BucketWithVersioningSSEncryptionAndAccessLogging`
- the `CreateS3BucketWithVersioningSSEncryptionAndAccessLogging` func is also only called from 1 more place - the `createS3BucketWithRetry` method and following the chain of functioncalls, only when `func (remoteState *RemoteState) Initialize(...)` is called.
- based on the above 2 observations, I'm not following exactly where the concern that this behaviour will impact existing users will come from. As far as I see, this behaviour of `DoesS3BucketExist` has been pushed closer to the actual s3 client API call to create the bucket, but it doesn't ultimately change the behaviour we execute when calling into `terragrunt apply | plan | validate`
- we do get the prompt every time to create the bucket, even if `terragrunt validate` has been called - the user should still be prompted to see if they want the S3 bucket created or not

Comment (ina-stoyanova) [reply remote/remote_state_s3.go:346]: I agree that this change can be extracted into its own separate PR, so I'll raise a separate one for it, so we can have an appropriate discussion.

File: remote/remote_state_s3.go
Comment (ina-stoyanova) [reply remote/remote_state_s3.go:428]: Hm, I'm unsure what you meant here

File: remote/remote_state_s3.go
Comment (ina-stoyanova) [reply remote/remote_state_s3.go:426]: did you mean something like in the method being called (See lines 654-656). I think it's a fair point though, as it will read easier if the logging as all done at the same level.

File: remote/remote_state_s3.go
Comment (ina-stoyanova) [reply remote/remote_state_s3.go:428]: ahh. ignore the above ☝️ I _think_ I understand your train of thought: the error message is about the creation of the bucket, not the access logging 👍 Let me know if my understadning is correct.

File: remote/remote_state_s3.go
Comment (yorinasub17) [reply remote/remote_state_s3.go:346]: The concern I have is that `CreateS3BucketWithVersioningSSEncryptionAndAccessLogging` will do additional mutating actions to the bucket after `CreateS3Bucket` is called, specifically:

- `EnableEnforcedTLSAccesstoS3Bucket`
- `EnablePublicAccessBlockingForS3Bucket`
- `TagS3Bucket`
- `EnableVersioningForS3Bucket`
- `EnableSSEForS3BucketWide`
- `EnableAccessLoggingForS3BucketWide`

Before, this whole routine was skipped when the bucket already existed, meaning that we did not attempt to mutate that bucket to enable these features. Now, we are only skipping the create routine, meaning that we will attempt to do these mutations on that existing bucket, which may have been created before all of this code existed. My main concern here is that we may add further mutations in this routine (like we are doing now in this PR), and we will need to make sure that those mutations work even if the bucket was created in older versions of terragrunt, potentially as far back as `0.18`.

Comment (ina-stoyanova) [reply remote/remote_state_s3.go:346]: ahaa! got it - yeah, totally missed this! Ok, just about to raise the separate PR for this.

File: remote/remote_state_s3.go
Comment (yorinasub17) [reply remote/remote_state_s3.go]: Yup that is correct. E.g., I would switch this to `("Error creating access logging bucket: %s", err)`.

File: remote/remote_state_s3.go
Comment (yorinasub17) [reply remote/remote_state_s3.go:426]: No I mean that since this is doing multiple steps, it would help to have a log message here indicating the two steps it is doing (creating access logging bucket and then hooking it up). I think that is slightly better in terms of debuggability.

It's not a big deal though so if you disagree that is helpful, feel free to ignore (hence labeled as `nit`).

Comment (ina-stoyanova) [reply remote/remote_state_s3.go:439]: ah I see! that's a good suggestion indeed! will get it in there :)

**Comment by ina-stoyanova** [comment]
I've not also gone ahead and added a new integration test as suggested @yorinasub17. This is still [WIP] as I needed to touch `terratest` as such functionality wasn't there yet. [Please see this Draft PR. ](https://github.com/gruntwork-io/terratest/pull/689)

I'm worrying that by touching terratest as part of this, it might be a bit of a scope creep, but the changes seem quite quick anyway. What's your take on this?

**Comment by yorinasub17** [comment]
> What's your take on this?

You can always start by having the function as a private function in the repo that will use it (in this case `terragrunt`), and then add a comment or file a TODO to port it to terratest. Even better if you start the PR in terratest (like you did here) so that we are less likely to forget it.

The main goal here is to disassociate merging this in from the terratest PR so that they can be released at different times. That way, if you end up in a massive yak shave to generalize the function or we have extended conversations about whether or not the function makes sense in terratest, it doesn't block this PR.

**Comment by ina-stoyanova** [comment]
@yorinasub17 Have a look at the issue I've created [here](https://github.com/gruntwork-io/terragrunt/issues/1404) for the purposes of the discussion in [your comment ](https://github.com/gruntwork-io/terragrunt/pull/1400#discussion_r513017599)

File: remote/remote_state_s3.go
Comment (ina-stoyanova) [reply remote/remote_state_s3.go]: > This change of moving `DoesS3BucketExist` check to `CreateS3Bucket` makes me nervous because it is a fairly significant change in behavior.
> 
> Before, terragrunt never attempted to "migrate" an existing bucket to match the configurations and defaults that you have set. This means that even if we update bucket parameters across releases, and if the user upgrades terragrunt versions, the s3 bucket remained untouched. This is somewhat surprising and confusing behavior for folks who are used to terraform, but has a few advantages for us. Off the top of my head:
> 
> * It is simpler to maintain, as we didn't really have to worry about backwards compatibility in the s3 bucket creation code: we always started from scratch with a new bucket.
> * It doesn't have the behavior of modifying the state bucket during a `plan` or `validate` stage, which can be surprising to a user attempting to do a "read-only" task (NOTE: creating the bucket also happens during these stages, which is arguably just as surprising, but in the create case there is an interactive prompt to confirm the creation that softens the impact).
> 
> There is potential that this change can break a lot of existing users, and we also should think through the implications of applying some of the updates that this will do. From the looks of it, this change appears to be orthogonal to the access logging change, so I think this specific update should be in a separate PR where we can discuss the implications of this update.



> here is to disassociate merging this in

So, I just pushed the last bit of changes i've added to this repo, but so far the test is not proving anything, as it points to  the not-implemented function in Terratest yet. 

Do you suggest that I implement a temp function using the s3client here, or is leaving it as a commented `TODO` like you've pointed out enough?

**Comment by ina-stoyanova** [comment]
> > What's your take on this?
> 
> You can always start by having the function as a private function in the repo that will use it (in this case `terragrunt`), and then add a comment or file a TODO to port it to terratest. Even better if you start the PR in terratest (like you did here) so that we are less likely to forget it.
> 
> The main goal here is to disassociate merging this in from the terratest PR so that they can be released at different times. That way, if you end up in a massive yak shave to generalize the function or we have extended conversations about whether or not the function makes sense in terratest, it doesn't block this PR.

Roger that! Thanks for the support!

So, I just pushed the last bit of changes I've added to this repo, but so far the test is not proving anything, as it points to the not yet implemented function in Terratest.

Do you suggest that I implement a temp function using the s3client in terragrunt & get the test passing, or is leaving it as a commented TODO like you've pointed out enough?

File: remote/remote_state_s3.go
Comment (yorinasub17) [inline remote/remote_state_s3.go:443]: NIT: Consider including the access logging bucket name in the log message.

File: remote/remote_state_s3.go
Comment (yorinasub17) [inline remote/remote_state_s3.go:448]: NIT: consider adding a log message here, similar to above for the create error.

File: test/integration_test.go
Comment (yorinasub17) [inline test/integration_test.go:609]: NIT:
```suggestion
	s3BucketLogsName := fmt.Sprintf("%s-tf-state-logs", s3BucketName)
```

**Review by yorinasub17 — COMMENTED** [review]
Updates LGTM! Found a few more minor nits, but otherwise the changes makes sense to me!

File: docs/_docs/04_reference/config-blocks-and-attributes.md
Comment (brikis98) [inline docs/_docs/04_reference/config-blocks-and-attributes.md:241]: NIT:

```suggestion
- `skip_bucket_accesslogging`: _DEPRECATED_ If provided, will be ignored. A log warning will be issued in the console output to notify the user.
```

File: docs/_docs/04_reference/config-blocks-and-attributes.md
Comment (brikis98) [inline docs/_docs/04_reference/config-blocks-and-attributes.md:250]:
```suggestion
- `accesslogging_bucket_name`: (Optional) When provided as a valid `string`, create an S3 bucket with this name to store the access logs for the S3 bucket used to store Terraform state. If not provided, or string is empty or invalid S3 bucket name, then server access logging for the S3 bucket storing the terraform state will be disabled.
```

Also, shouldn't empty string or invalid bucket name or any other invalid input produce an error, rather than just silently disabling access logging?

File: remote/remote_state_s3.go
Comment (brikis98) [inline remote/remote_state_s3.go:401]: Apologies if I missed the discussion around this, but why remove the wait?

File: remote/remote_state_s3.go
Comment (brikis98) [inline remote/remote_state_s3.go:469]: NIT:

```suggestion
			return CreateS3Bucket(s3Client, logsBucketName, terragruntOptions)
```

File: test/integration_test.go
Comment (brikis98) [inline test/integration_test.go:602]: Looks like that Terratest PR is still in draft state. Is that intentional? Also, you might want to put a TODO above this to make reduce the odds of us forgetting about this!

**Review by brikis98 — COMMENTED** [review]
Very close! Just some minor comments.

File: remote/remote_state_s3.go
Comment (ina-stoyanova) [reply remote/remote_state_s3.go:401]: Good spot, Jim! I meant to comment on this, and confirm this is a duplicated piece of logic. Apologies for forgetting to do so earlier. 

I've removed it thinking it was a duplicate. See commit https://github.com/gruntwork-io/terragrunt/pull/1400/commits/e77cde2ecef9eb2f576c6c6e825d7d10edafb157

File: docs/_docs/04_reference/config-blocks-and-attributes.md
Comment (ina-stoyanova) [reply docs/_docs/04_reference/config-blocks-and-attributes.md:250]: So any invalid input for an S3 bucket name will produce an error from the S3 client: `InvalidBucketName: The specified bucket is not valid.`.

I've decided to leave the case where "empty" because, I didn't know how to differentiate between `""` and `nil` for strings - they seem to be treated the same coming from the config. 

One option is as you've suggested to treat these two cases separately (_maybe with checking the length of the string passed_), and/or update the logs - and be specific why the logging was disabled.

File: test/integration_test.go
Comment (ina-stoyanova) [reply test/integration_test.go:602]: The terratest PR was failing the tests, so wanted to see why they were failing. Will wrap it up as well, once I'm happy the tests are going through

File: remote/remote_state_s3.go
Comment (brikis98) [reply remote/remote_state_s3.go:401]: Why do you think it's duplicated?

Comment (ina-stoyanova) [reply remote/remote_state_s3.go:401]: _Actually, after re-reading it for the 5th time, I think I was massively misreading what the flow of the logic is 🤦 ._

_What I was thinking:_ It's already handling the error & waiting for the bucket to exist.= in the guard block (line 395-401)

_What's I think is actually happening. P;ease correct me if I'm wrong in this :) _ It's only surpressing the 'OwnedByYouError' so it doesn't exit prematurely & waits instead. But if there's no error thrownm we still want to make sure the bucket has been created fine - hence the second `WaitUntilS3...`

Comment (brikis98) [reply remote/remote_state_s3.go:401]: Yea, the check on 395-401 is an extra check for concurrent bucket creation: e.g., if you run `terragrunt apply-all`, it will execute on a bunch of modules in parallel, all of which might try to create the exact same bucket at the same time.

The check `WaitUntilS3BucketExists` I highlighted in this comment ensures that _after_ the bucket has been created, we wait long enough that we can "see" it in the AWS APIs. This is to work around the fact that AWS APIs are async and eventually consistent. Without this check, the code would move to the following parts, which try to modify the S3 bucket, and potentially hit errors because the bucket information hasn't propagated yet.

Comment (ina-stoyanova) [reply remote/remote_state_s3.go:401]: you're absolutely right - thanks for spotting & explaining 🎈 
*Reverting this commit now*

**Comment by ina-stoyanova** [comment]
Some new commits here. Another review would be very welcome 👀 🙂

**Review by brikis98 — APPROVED** [review]
LGTM! 

Note: I'm not sure if you're waiting to merge the Terratest thing first or merging this and coming back to update the Terratest version later?

**Comment by ina-stoyanova** [comment]
Thank you @brikis98 🎉 

I think the reasoning we agreed on with @yorinasub17 was to not necessarily wait on the terratest, as that might be a yak shave, as it's turning out. I'll release this first thing on Monday, and then keep as a priority to close the terratest PR too.

**Pull request merged**
